main.py

```diff
@@ -1,30 +1,91 @@
+        os.makedirs(os.path.dirname("./out/"), exist_ok=True)
+        if(grab(url, "./out/")):
+            for file in os.listdir('./out/'):
+                if fnmatch.fnmatch(file, '*.json'):
+                   template_path = os.path.join('./out/', file)
+                   os.rename(template_path, os.path.join('./out/', 'template.json'))
+                   return True
+        else:
+            parsed_url = urlparse(url)
+            basename = os.path.basename(parsed_url.path)
+            title, ext = os.path.splitext(basename)
+            if not title:
+                title = "template"
+            if not ext:
+                ext = "json"
+            with requests.get(url, stream=True, timeout=10) as resp:
+                if resp.status_code == 200:
+                    with open("./out/template.json", "wb") as json_file:
+                        for chunk in resp.iter_content(chunk_size=8192):
+                            json_file.write(chunk)
+                    return True
+                else:
+                    return False
+           
+    def template_based_extract(self, url):
+        if(self.fetch_template(url)):
+            with zipfile.ZipFile("export.zip", mode="w") as archive:
+                for document in self.get_documents():
+                    with open("file.pdf", "wb") as pdf_file:
+                        pdf_file.write(document.pdf)
+                    data_frame_list = tabula.read_pdf_with_template("./file.pdf", "./out/template.json") 
+                    # Tabula's read_pdf_with_template() returns a list of data frames we can append to form a CSV. 
+                    for data_frame in data_frame_list:
+                        data_frame.to_csv(f"{document.slug}.csv", mode="a", index=False, header=False)
+                    archive.write(f"{document.slug}.csv")
+        else:
+            self.template_less_extract()
+    
+    def template_less_extract(self):
+        with zipfile.ZipFile("export.zip", mode="w") as archive:
+            for document in self.get_documents():
+                with open("file.pdf", "wb") as pdf_file:
+                    pdf_file.write(document.pdf)
+                tabula.convert_into("file.pdf", f"{document.slug}.csv", output_format="csv", pages="all",) 
+                # Tabula's convert_into() guesses boundaries for table extraction. 
+                archive.write(f"{document.slug}.csv")
 
-        with open(f"{document.slug}.csv") as csv_file:
-            self.upload_file(csv_file)
+    def main(self):
+        """
+        If a template is provided, it extracts dataframes from the PDF(s)
+        using that template and appends it to a CSV for each document and returns
+        a zip file of all the CSVs. If no template is provided, it guesses
+        the boundaries for each file
+        """
+        try:
+            url = self.data["url"]
+        except Exception as exc:
```

Comment:
You should *almost* never use an `except Exception` as it can cover errors you have not accounted for.  In this case, the only exception should be a `KeyError`, but this is written more idiomatically as:
```
url = self.data.get("url")
```
If `url` is not set, it returns None.  If you wanted a different default, you can set it as the second argument to get, which can be useful.